server/common/EntryParser.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (C) 2021 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.server.common;
+
+import com.vaticle.typedb.common.yaml.Yaml;
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+import com.vaticle.typedb.core.server.common.CommandLine.Option;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import static com.vaticle.typedb.common.util.Objects.className;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_CAST;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.CONFIG_OPTION_ENUM_UNEXPECTED_VALUE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.CONFIG_OPTION_UNEXPECTED_VALUE_TYPE;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Server.MISSING_CONFIGURATION_OPTION;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+import static com.vaticle.typedb.core.server.common.Util.scopeOption;
+
+public abstract class EntryParser implements Option.CliHelp {
+
+    private final String description;
+
+    EntryParser(String description) {
+        this.description = description;
+    }
+
+    String description() {
+        return description;
+    }
+
+    static abstract class ValueParser<TYPE> {
+
+        abstract TYPE parse(Yaml yaml, String scope);
+
+        boolean isLeaf() {
+            return false;
+        }
+
+        Leaf<TYPE> asLeaf() {
+            throw TypeDBException.of(ILLEGAL_CAST, className(getClass()), className(Leaf.class));
+        }
+
+        boolean isNested() {
+            return false;
+        }
+
+        Nested<TYPE> asNested() {
+            throw TypeDBException.of(ILLEGAL_CAST, className(getClass()), className(Nested.class));
+        }
+
+        static abstract class Nested<T> extends ValueParser<T> {
+
+            abstract List<Help> help(String scope);
+
+            @Override
+            boolean isNested() {
+                return true;
+            }
+
+            @Override
+            Nested<T> asNested() {
+                return this;
+            }
+        }
+
+        static class Leaf<T> extends ValueParser<T> {
```

Comment:
we pre-define some types of leaves/primitive value parsers. We could also do things like define a `Path` type here!